### PR TITLE
KEYCLOAK-12615 HS384 and HS512 support for Client Authentication by Client Secret Signed JWT

### PR DESCRIPTION
**Pull request by tnorimat**
JIRA issue is as follows.
https://issues.jboss.org/browse/KEYCLOAK-12615

[How to determine the signature algorithm]
Unlike the JWT generated and signed by keycloak, There is the case that the JWT generated and signed by the client does not include "alg" claim because JWK RFC states that "alg" claim is OPTIONAL.
(https://tools.ietf.org/html/rfc7517#section-4.4)

However, "kty" claim is MUST.
(https://tools.ietf.org/html/rfc7517#section-4.1)

Considering these points, try to determine "alg" as follows if "alg" does not exist.
 "kty" = "RSA" -> "alg" = "RS256" (backward compatibility : the previous implementation on treating Client signed JWT hardcodes RS256)
 "kty" = "EC"  -> "alg" = "ES256" 
 "kty" = "oct" -> "alg" = "HS256" (backward compatibility : the previous implementation on treating Client signed JWT hardcodes HS256)
 "kty" does not exists -> return null (error)
 "kty" = other value -> return null (error)
referring JWA RFC https://tools.ietf.org/html/rfc7518#section-6.1

As for KeyType.EC, JWA RFC states that "crv" is MUST.
(https://tools.ietf.org/html/rfc7518#section-6.2.1.1)
It would be desirable that keycloak user "crv" and determine the algorithm.
However, current JWSHeader implementation can not provides "crv" value. So use ES256.

As for KeyType.OCT, keycload defines it as upper case but JWA RFC states lower case
(https://tools.ietf.org/html/rfc7518#section-6.1)
For backward compatibility, use String.equalsIgnoreCase() not String.equals().

I'm not sure but it seems that this PR also resolve https://issues.redhat.com/browse/KEYCLOAK-12299 .

### CODE REVIEW

**Review by mposolda — CHANGES_REQUESTED** [review]
@tnorimat Thanks a lot for the PR!
Few points:
- I left one inline comment in the adapter support class JWTCLientSecretCredentialProvider. It will be also good to add docs for that one as I pointed in the inline comment?
- The other PR for https://issues.redhat.com/browse/KEYCLOAK-12299 was actually merged in the meantime. So I suggest to rebase your PR and doublecheck that your PR is not somehow in conflict with that. Or maybe even revert changes done by that PR if you think that your PR does the same thing in more robust way etc. Leaving to you if something needs to be done or not :)

Thanks

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authentication/JWTClientSecretCredentialsProvider.java [hunk 1]
```diff
@@ -11,70 +11,87 @@
...
-   
+
     public void setClientSecret(String clientSecretString) {
         // Get client secret and validate signature
         // According to <a href="http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication">OIDC's client authentication spec</a>,
         // The HMAC (Hash-based Message Authentication Code) is calculated using the octets of the UTF-8 representation of the client_secret as the shared key. 
         // Use "HmacSHA256" consulting <a href="https://docs.oracle.com/javase/jp/8/docs/api/javax/crypto/Mac.html">java8 api</a>
         // because it must be implemented in every java platform.
-        clientSecret = new SecretKeySpec(clientSecretString.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
+        setClientSecret(clientSecretString, Algorithm.HS256);
```
[the remark this answers is not in the document]
Comment (tnorimat) [reply adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authentication/JWTClientSecretCredentialsProvider.java]: I see, as for client adapter's new algorithm support about Client Signed JWT, I'll add codes treating them and also add documentation about it.

**Comment by tnorimat** [comment]
As for the other PR for https://issues.redhat.com/browse/KEYCLOAK-12299 you pointed out, this PR does not resolve the problem https://github.com/keycloak/keycloak/pull/6640 . It's my mistake.

While https://github.com/keycloak/keycloak/pull/6640 treats the JWK obtained from jwks_uri the client application hosts, this PR's codes treat the JOSE header of Client Signed JWS. I've been confused about them (JWK and JOSE header). 

Also, I've found that some part of this PR's codes (checking whether "alg" parameter exists and the fallback if not exist) are not needed because the JOSE header's specification (https://tools.ietf.org/html/rfc7515#section-4.1.1) tells that "alg" parameter MUST be presented. Therefore, I'll remove these codes.

**Comment by mposolda** [comment]
@tnorimat Thanks! Let me know when the PR is ready for re-review.

**Comment by tnorimat** [comment]
I've revised this PR. Also, I've sent the document PR (https://github.com/keycloak/keycloak-community/pull/44).

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authentication/JWTClientSecretCredentialsProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mposolda) [reply adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authentication/JWTClientSecretCredentialsProvider.java]: Thanks!

**Review by mposolda — APPROVED** [review]
@tnorimat Thanks! Approving. Should be ready once docs PR is reviewed too.

**Pull request merged**